src/main/kotlin/io/emeraldpay/dshackle/upstream/AbstractHead.kt

```diff
@@ -50,25 +56,28 @@ abstract class AbstractHead @JvmOverloads constructor(
     private var stopping = false
     private var lastHeadUpdated = 0L
     private val lock = ReentrantLock()
+    private var future: Future<*>? = null
+    private val delayed = AtomicBoolean(false)
 
     init {
-        val state = AtomicBoolean(false)
-        Gauge.builder("stuck_head", state) {
+
+        val className = this.javaClass.simpleName
+
+        Gauge.builder("stuck_head", delayed) {
             if (it.get()) 1.0 else 0.0
-        }.tag("upstream", upstreamId).tag("class", this.javaClass.simpleName).register(Metrics.globalRegistry)
+        }.tag("upstream", upstreamId).tag("class", className).register(Metrics.globalRegistry)
         Gauge.builder("current_head", forkChoice) {
             it.getHead()?.height?.toDouble() ?: 0.0
-        }.tag("upstream", upstreamId).tag("class", this.javaClass.simpleName).register(Metrics.globalRegistry)
-        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(
-            {
-                val delay = System.currentTimeMillis() - lastHeadUpdated
-                val delayed = delay > awaitHeadTimeoutMs
-                state.set(delayed)
-                if (delayed) {
-                    log.warn("No head updates $upstreamId for $delay ms @ ${this.javaClass}")
-                }
-            }, 300, 30, TimeUnit.SECONDS
-        )
+        }.tag("upstream", upstreamId).tag("class", className).register(Metrics.globalRegistry)
+
+        instances.computeIfAbsent(className) {
+            AtomicInteger(0).also { toHeadCountMetric(it, "allocated") }
+        }.incrementAndGet()
+    }
+
+    protected fun finalize() {
```

Comment:
I think we should never use finalize because it forces jvm move that object to old-gen.